### PR TITLE
Refactor Elegy using Treex

### PR DESCRIPTION
**Pull request by cgarciae**
Elegy is now based on [Treex](https://github.com/cgarciae/treex) 🎉

## Changes
* Remove the `module`, `nn`, `metrics`, and `losses` from Elegy, instead Elegy reexports these modules from Treex.
* `GeneralizedModule` and friends are gone, to use Flax Modules use the `elegy.nn.FlaxModule` wrapper. 
* Low level API is massively simplified:
  * `States` is removed, since `Model` is a pytree all parameters are tracked automatically thanks to Treex / Treeo.
  *  All static state arguments (`training`, `initializing`) are removed, `Module`s can simply use `self.training` to pick their training state and `self.initializing()` to check whether they are initializing.
  * Signature for `pred_step`, `test_step`, and `train_step` now simply consists of `inputs` and `labels`, where `labels` is a `dict` that can contain additional keys like `sample_weight` or `class_weight` as required by the losses and metrics.

### CODE REVIEW

**Comment by cgarciae** [comment]
@alexander-g This version uses a new Module system + layers from Treex and hooks are gone. 
Resnets and the GAN example needs fixing if you want to give it a try.

**Comment by cgarciae** [comment]
Will merge this for now. Docs need updating.
